config_parser.cc

```diff
@@ -199,14 +201,22 @@ bool NginxConfigParser::Parse(std::istream* config_file, NginxConfig* config) {
           new_config);
       config_stack.push(new_config);
     } else if (token_type == TOKEN_TYPE_END_BLOCK) {
-      if (last_token_type != TOKEN_TYPE_STATEMENT_END) {
+       open_count--;
+       if (last_token_type != TOKEN_TYPE_STATEMENT_END) {
         // Error.
-        break;
-      }
+        if (last_token_type != TOKEN_TYPE_END_BLOCK &&
+               last_token_type != TOKEN_TYPE_START_BLOCK) {
+             break;
+        }
+        else if (open_count < 0) {
```

Comment:
Logically, all looks good! You could maybe check this right after the decrement on 204, but that's a stylistic choice. You can merge